### PR TITLE
Fix permissions for items shown in admin dashboard

### LINKED ISSUE #1641: I added a new user with all permissions but that user can't make new users
Problem in title. Is that just the default behavior that only the original admin can create new accounts?

### PR DESCRIPTION
**Pull request by impactmass**
First part of the resolution of #1651.
Also resolve #1641. 

**What this PR solves:**
- When owner gives dashboard access to an account, the account has access to everything manageable in the dashboard. This PR fixes that. So, only the packages that the account has access to shows up in the dashboard.

**To test:**
- Run the app on this branch.
- Create an account other than admin account (you can use incognito or another browser... so you have both the admin and the new account opened)
- Grant dashboard access to that account (i.e toggle on "Dashboard" in Accounts -> Manage for that user)
- Go through items showing in the dashboard for that account
- Confirm that not all available settings show up in the dashboard for that account.
- Also confirm that publish buttons/controls don't show up at the top bar (except with access to "createProduct")... [see screenshot](https://cloud.githubusercontent.com/assets/439959/25412459/b41c7a3e-29d7-11e7-8c18-73059ad90387.png)
- Toggle on and off dashboard items for this account from your own admin account, then go to the dashboard for the test account to see if access is being given accordingly.

### CODE REVIEW

File: imports/plugins/core/dashboard/client/containers/packageListContainer.js [hunk 1]
```diff
@@ -15,9 +15,9 @@ function handleShowPackage(event, app) {
 }
 
 function composer(props, onData) {
-  const settings = Reaction.Apps({ provides: "settings", enabled: true }) || [];
-
-  const dashboard = Reaction.Apps({ provides: "dashboard", enabled: true })
+  const roles = Roles.getRolesForUser(Meteor.userId(), Reaction.getShopId());
```
Comment (mikemurray) [inline imports/plugins/core/dashboard/client/containers/packageListContainer.js]: Maybe this should be move into the `Reacton.Apps` method. so any method that calls reaction apps without an `audience` will automatically be filtered.

**Comment by impactmass** [comment]
Hi @mikemurray, please take a look at the diff for how I selectively showed the icons in the toolbar based on permission (i.e in `toolbar.js` & `toolbarContainer.js`). I'd like comments on the direction I took there.

**Comment by spencern** [comment]
@impactmass 
Merging this branch into Marketplace leaves accounts with shop admin permissions for secondary / marketplace shops without any profile or dashboard links at all.

Here's this users permission set
```js
	"roles" : {
		"J8Bhq3uTtdgwZx3rz" : [
			"account/profile",
			"guest",
			"product",
			"tag",
			"index",
			"cart/checkout",
			"cart/completed",
			"notifications"
		],
		"KQ9dY2qPNuTjsb2EL" : [
			"admin",
			"seller",
			"guest",
			"manage-users",
			"orders",
			"account/profile",
			"product",
			"createProduct",
			"tag",
			"index",
			"cart/checkout",
			"cart/completed",
			"stripe-connect-redirect",
			"test-role",
			"test-role33"
		]
	}
```

### User Dropdown
image

### Dashboard
image

### Edit Mode
image

Edit mode appears to be on, because you can see the "edit tags" link and clicking into that opens the tag editor
image

But there is no toggle for edit mode either.

**Comment by impactmass** [comment]
Thanks, addressing these now.

**Comment by impactmass** [comment]
@spencern 
response to the issues:

1. **Broken User Dropdown & Dashboard**: When testing/merging this branch with Marketplace, we'll need to change `getShopId()` to `getSellerShopId()` in some places (so that the default shop is not used everytime. The lines specific to this issue are [here](https://github.com/reactioncommerce/reaction/blob/3e21aeb4362613fab1b17bf8f5286362862681aa/client/modules/core/helpers/apps.js#L65), [here](https://github.com/reactioncommerce/reaction/blob/3e21aeb4362613fab1b17bf8f5286362862681aa/client/modules/core/helpers/apps.js#L111), and [here](https://github.com/reactioncommerce/reaction/blob/3e21aeb4362613fab1b17bf8f5286362862681aa/imports/plugins/core/dashboard/client/containers/toolbarContainer.js#L90). 

Changing them reveals an issue: the new Package documents created for new shops do not ref the shopId, so a query by ReactionApps with this sample filter:
```
{
    "registry.provides": "userAccountDropdown",
    "enabled": true,
    "shopId": "RuTERn4A4nzZvKELo" // new seller shopId returned from getSellerShopId()
}
```
returns no packages.

Manually testing out `db.getCollection('Packages').find({"shopId": "RuTERn4A4nzZvKELo" })` returns 
returns no results. The default `J8Bhq3uTtdgwZx3rz` returns. 

I'm not sure where to debug this on the Marketplace branch. A thought is that maybe we work on fixing this on the Marketplace branch after this PR passes review into development. (?)

2. **Edit Mode:** Once the above is resolved, the edit mode should also work as expected; since `createProduct` permission which the visibility of those icons are based on, is included as part of seller permissions.

**Comment by spencern** [comment]
@impactmass Thanks for this great write up. Testing this stuff now.

**Comment by spencern** [comment]
Re: Packages returning no results. I'm not seeing that.

```js
meteor:PRIMARY> db.Shops.find({}, {_id: 1})
{ "_id" : "J8Bhq3uTtdgwZx3rz" } // Primary Shop
{ "_id" : "ddzuN2YPvgvx7rJS5" }
{ "_id" : "KQ9dY2qPNuTjsb2EL" } // Seller Shop
```
Primary Shop returns 47 packages
```js
meteor:PRIMARY> db.Packages.find({shopId: "J8Bhq3uTtdgwZx3rz"}).count()
47
```
Seller Shop returns 47 packages
```js
meteor:PRIMARY> db.Packages.find({shopId: "KQ9dY2qPNuTjsb2EL"}).count()
47
```

I do get results for each shop running this:
db.Packages.find({shopId: "KQ9dY2qPNuTjsb2EL", "registry.provides": "userAccountDropdown", enabled: true})


I think it's called a `shortcut` in the registry, e.g. `"registry.provides" : "shortcut"`

You can see this here in the Accounts plugin's register.js file here https://github.com/reactioncommerce/reaction/blob/development/imports/plugins/core/accounts/register.js#L32

I _think_ but @mikemurray might know better, that `userAccountsDropdown` is used for users without permissions (and currently is only the user profile)

**Comment by impactmass** [comment]
OK, thanks.
I'm going to tripple check this, but this is my screen while double checking what I said earlier: https://www.dropbox.com/s/1tw9eibedtld1hn/mkpl.mov?dl=0 (no results).

**Comment by spencern** [comment]
Hmm, I have restarted the app since checking on this, let me see if that changes things

**Comment by spencern** [comment]
Ok, so after a bit of digging I think I've found the issue here:

It  appears that packages for newly created shops don't get added to the database until after a server restart.

```js
meteor:PRIMARY> db.Shops.find().count()
2
meteor:PRIMARY> db.Packages.find().count()
94
meteor:PRIMARY> db.Shops.find({}, {_id: 1})
{ "_id" : "J8Bhq3uTtdgwZx3rz" }
{ "_id" : "ddzuN2YPvgvx7rJS5" }

// Create new shop

meteor:PRIMARY> db.Shops.find({}, {_id: 1})
{ "_id" : "J8Bhq3uTtdgwZx3rz" }
{ "_id" : "ddzuN2YPvgvx7rJS5" }
{ "_id" : "zPtjvoe896GPTcA9y" }
meteor:PRIMARY> db.Packages.find().count()
94 // <= only includes the original two shop's worth of Packages

// SERVER RESTART

meteor:PRIMARY> db.Shops.find({}, {_id: 1})
{ "_id" : "J8Bhq3uTtdgwZx3rz" }
{ "_id" : "ddzuN2YPvgvx7rJS5" }
{ "_id" : "zPtjvoe896GPTcA9y" } // <= New Shop

meteor:PRIMARY> db.Packages.find().count()
141 // <= Includes new shop's packages now
```

**Comment by impactmass** [comment]
Final checks on this please:

- @mikemurray please confirm the approach on permission done on toolbar component
- @zenweasel please confirm PR works according to steps described
- @spencern after the issue we figured out yesterday, when I merge this PR into `markeplace` and made the modifications of `shopId` in ReactionApps & Package publication, things worked fine (fine). I can't put those changes in this PR, since this is going into development (we'll have to resolve those in marketplace)

If all is fine please approve. If any issues please comment. I'll address as fast as possible to speed things up.

File: imports/plugins/core/dashboard/client/components/toolbar.js [hunk 2]
```diff
@@ -113,7 +122,7 @@ class PublishControls extends Component {
   }
 
   renderCustomControls() {
-    if (this.props.dashboardHeaderTemplate) {
+    if (this.props.dashboardHeaderTemplate && this.props.hasCreateProductAccess) {
```
Comment (spencern) [inline imports/plugins/core/dashboard/client/components/toolbar.js]: Not specific to this PR, but has the `createProduct` permission become a bit of a universal check for more than just creating products?

**Comment by kieckhafer** [comment]
Does what it's supposed to do. However, I have one comment on how this works. The `Dashboard` setting needs to be enabled in order to see the dashboard, however you can enable all the other settings first, and leave the Dashboard off, meaning a person has all these permissions, but can't access them until the Dashboard switch is flipped.

One place in particular this is an issue is with the Profile dropdown, where you can see settings available for clicking, but when you click them, they don't do anything unless the Dashboard setting is enabled to.

For example, If I just have the "Orders" permissions granted, but not the dashboard, Orders will show up in the account dropdown, but clicking on it does nothing.

Maybe we need to wrap the profile dropdown with a `isDashboardEnabled` setting as well?

**Comment by impactmass** [comment]
Thanks for testing this out. and those are good point(s) @kieckhafer.

And like you said, #1651

**Comment by impactmass** [comment]
Thanks for testing this out. and those are good point(s) @kieckhafer.

And like you said, there's still more to be done around the permissions. 

I see the ticket #1651 in [3 parts](https://github.com/reactioncommerce/reaction/issues/1651#issuecomment-297111475). This PR handles the first part of restricting what shows up in the dashboard based on access levels. 

The second part (which I haven't gotten on yet) can handle this weirdness that you described. We should have a discussion around approaches for that sometime.

But I think we can merge this in for now (as a first step).

**Comment by mikemurray** [comment]
@impactmass toolbar permissions looks good to me

**Pull request merged**